bftengine/src/preprocessor/PreProcessor.hpp

```diff
@@ -47,16 +48,56 @@ struct RequestState {
   uint64_t reqRetryId = 1;
 };
 
-// Pre-allocated (clientId * dataSize) buffers
-typedef std::deque<std::pair<std::atomic_bool, concordUtils::Sliver>> PreProcessResultBuffers;
-typedef std::shared_ptr<RequestState> RequestStateSharedPtr;
-// (clientId * dataSize + reqOffsetInBatch) -> RequestStateSharedPtr
-typedef std::unordered_map<uint16_t, RequestStateSharedPtr> OngoingReqMap;
+using RequestStateSharedPtr = std::shared_ptr<RequestState>;
+
+//**************** Class RequestsBatch ****************//
+class PreProcessor;
 
+class RequestsBatch {
+ public:
+  RequestsBatch(PreProcessor &preProcessor, uint16_t clientId) : preProcessor_(preProcessor), clientId_(clientId) {}
+  void init();
+  void registerBatch(const std::string &cid, uint32_t batchSize);
+  void startBatch(const std::string &cid, uint32_t batchSize);
+  void addReply(PreProcessReplyMsgSharedPtr replyMsg);
+  bool isBatchRegistered() const { return batchRegistered_; }
+  bool isBatchInProcess() const { return batchInProcess_; }
+  void increaseNumOfCompletedReqs() { numOfCompletedReqs_++; }
+  RequestStateSharedPtr &getRequestState(uint16_t reqOffsetInBatch);
+  const std::string getCid() const;
+  void cancelBatch();
+  void releaseReqsAndSendBatchedReplyIfCompleted();
+  void finalizeBatchIfCompleted();
+  void handlePossiblyExpiredRequests();
+  void cancelBatchedPreProcessingOnNonPrimary(const ClientMsgsList &clientMsgs, NodeIdType destId);
+
+ private:
+  void setBatchParameters(const std::string &cid, uint32_t batchSize);
+  void resetBatchParams();
+
+ private:
+  PreProcessor &preProcessor_;
+  const uint16_t clientId_;
+  std::string cid_;
+  std::atomic_uint32_t batchSize_ = 0;
+  std::atomic_bool batchRegistered_ = false;
+  std::atomic_bool batchInProcess_ = false;
+  std::atomic_uint32_t numOfCompletedReqs_ = 0;
+  // pre-allocated map; request offset in batch -> request state
+  std::map<uint16_t, RequestStateSharedPtr> requestsMap_;
```

Comment:
As far as I understand this map is initialised once and later it is used only for reading. Maybe it can be initialised in the constructor and made const? 
Or at least a comment that it is intended for RO use and it is not protected with a mutex.

Reply:
NP, will add in the next MR.